### ISSUE TITLE
ColorAndAlpha

### ISSUE THREAD

**Issue by Justinfront** [azrafe7/hxPixels#2]
Azrafe7

If they are correct these would seem useful...

```
public function setPixelColorAndAlpha( x: Int, y: Int, col: Int, alpha: Int ) {
    var pos = (y * this.width + x) << 2;

    var a = alpha;
    var r = (col >> 16) & 0xFF;
    var g = (col >> 8) & 0xFF;
    var b = (col) & 0xFF;

    this.bytes.set((pos + this.format.A), a);
    this.bytes.set((pos + this.format.R), r);
    this.bytes.set((pos + this.format.G), g);
    this.bytes.set((pos + this.format.B), b);

}

/** Fills the specified rect area, with `value` expressed in ARGB format. Doesn't do any bound checking. */
public function fillRectColorAndPixel(x:Int, y:Int, width:Int, height:Int, col:Int, alpha: Int):Void {
    var pos = (y * this.width + x) << 2;

    var stridePixels = new Pixels(width, 1, true);
    stridePixels.format = this.format;
    var stride = width << 2;

    for (x in 0...width) stridePixels.setPixelColorAndPixel(x, 0, col, alpha );
    for (y in 0...height) {
        this.bytes.blit(pos, stridePixels.bytes, 0, stride);
        pos += this.width << 2;
    }
}
```

Also for consistency maybe fillRect should be renamed fillRect32 and a fillRect that just takes RGB be added?  Just often a user may have alpha separate from the solid color.

**Comment by azrafe7** [comment]
Nah, although I agree they're convenient, it's trivial for the user to separate alpha and solid color and then use the current API.

**Comment by Justinfront** [comment]
they are pretty useful I think for hxDaedalus since often color and alpha are separate let me explain my rather rough idea.
https://gist.github.com/Justinfront/43b9ef006a18f97c6499
Where I would push the PixelShapeEnum onto an array within the simple drawing implemention to allow redraw and mouse hittest, probably try to get something working and then worry about better approaches. 
I have been looking at porting some c code to perhaps make drawing lines with antialias and curves easier but still lots to do ran out of time so just a snap shot of progress.
https://gist.github.com/Justinfront/7ad8b882cf92e3170b88
So my current thoughts are to take SVG and process into a threshold bitmap in haxe c#
https://github.com/cambiata/HxSvg 
Then create a mesh and I should be able to even render paths to images without a front end.
At moment I am just expanding possibilies before ruling out concepts, but it would be ideal to take the backend only approach as far as possible.
But I think while it's trivial you can see that it's ideal to have these functions since later they may have optimisations, but better that is done in hxPixel and a interface that provides the functionality is provided the pixel class should worry about the details.
Maybe seeing my use you can see it to be useful but if not then no worries. If your good at c then I may have some questions on the c line porting.

**Comment by Justinfront** [comment]
absolutely nothing is tested and the pixel shapes don't yet have any of the antialias line code included yet, I can't rem if I have simple drawing class for the pixels yet, but you probably get my approach.

**Comment by azrafe7** [comment]
Yeah I get it... but separating them is just:

``` haxe
    var alpha = (value >> 24) & 0xFF;
    var solidColor = value & 0xFFFFFF;
```

And recently I've introduced some more convenience functions into hxPixels:
- you can use array access to manipulate specific bytes as if the data were in ARGB (so, f.e., using  `pixels[idx]` - where `idx` is a multiple of 4 - means to get/set the alpha value of the related pixel), without worrying about the underlying format
- when you use `getPixel/getPixel32()` you can easily access a single channel value (as in `var red = pixels.getPixel32(10, 10).R`)
- you can also do `var pixelValue:Pixel = 0x203040; pixelValue.A = 0xFF; image.setPixel32(10, 10, pixelValue);`

So, for the moment, I don't think adding those functions would help in any significant way (sure I could change my mind, and I'm open to discussion).

The links you posted are quite interesting! (reading more carefully through them ;)

By the way I've played with c in the past, so if you have any questions just shoot away and I'll try to be of help if possible (no guarantees though).

**Comment by azrafe7** [comment]
I had taken a closer look at http://members.chello.at/~easyfilter/bresenham.html, and maybe  `setPixelAA()` could be more or less written as

``` haxe
    public function setPixelAA(x:Int, y:Int, i:Int) {
        if (x < 0 || x >= pixels.width || y < 0 || y >= pixels.height) return;
        var a = 1 - i / 255;
        var backColor = pixels.getPixel32(x, y);
        var color:Pixel = 0xFF00FF00;
        color.A = Std.int((1 - a) * backColor.A + (a) * color.A);
        color.R = Std.int((1 - a) * backColor.R + (a) * color.R);
        color.G = Std.int((1 - a) * backColor.G + (a) * color.G);
        color.B = Std.int((1 - a) * backColor.B + (a) * color.B);
        pixels.setPixel32(x, y, color);
    }
```

**Comment by azrafe7** [comment]
Hey, I can happily give a hand tonight on porting the rasterizer (need to code something interesting), if you don't mind. 

Do you have a repo/gist for that, or would like me to look at a specific section?

(I've only ported plotLineAA and plotLineWidth for now)

**Comment by Justinfront** [comment]
azrafe7 all current progress is checked in, for all repositories it's the 'pixel' branch.

**Comment by azrafe7** [comment]
Ok! Going through it...

can you quick point me to something not yet implemented/tested while I check them one by one?

**Comment by azrafe7** [comment]
Oh, I see you commented them with the port status...

**Comment by Justinfront** [comment]
Ah that is not upto date, basically the ones in the code seem to work eg:
https://github.com/hxDaedalus/hxDaedalus/blob/pixel/src/wings/pixel/PixelCircle.hx
the plotCircle also works but the AA one as used looks nicer.
https://github.com/hxDaedalus/hxDaedalus/blob/pixel/src/wings/pixel/PixelLine.hx
the AA and width ones compile but are not good.
Currently I am not using any of the curve ones half of them crash or look bad instead I am using lots of lines but maybe this is cool since I may look to using mesh triangles for fill.

**Comment by Justinfront** [comment]
You may find the simple repository ideal for testing as you can check you don't break other stuff easily.

**Comment by Justinfront** [comment]
By the way I am looking at gif writing but not got far, I only have encoding LZW which is pretty basic but perhaps useful as I have not seen it implemented in Haxe elsewhere yet.  By using an 'add' function that you pass, it allows you to pass a wrapper to OutputBytes with writeString( Std.string(    or  add or push in the case of a List or an Array, still unsure on the format of the output required for gif but it is tested for List<Int> and OutputBytes.  I had something for Reading LZW but was not yet working and maybe not useful as already gif reading.
https://gist.github.com/Justinfront/d06f67d99957aa25872b
